### ISSUE TITLE
Beautify json before writing to config in cSpell

### ISSUE THREAD

**Issue by keaising** [jose-elias-alvarez/null-ls.nvim#1466]
### Issues

- [X] I have checked [existing issues](https://github.com/jose-elias-alvarez/null-ls.nvim/issues) and there are no existing ones with the same request.

### Feature description

[In existing code](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/lua/null-ls/builtins/code_actions/cspell.lua#L140), a raw json string is written to cspell.json without beautifying it.

Is there any way to beautify a json string (e.g. sort elements, keys in array and object, add proper indentation) before writing it to cspell.json?

In my research, I found that lua doesn't have a `JSON.stringify` function like js, so I think there are two ways to implement this feature:
1. embed another 3rd party library for beautifying
2. null-ls provide a callback after write to cspell.json, user can call some tools like `jq` to beautify config by theirselves

I have already read https://github.com/jose-elias-alvarez/null-ls.nvim/pull/1412#issuecomment-1454859416 and I just want to know this posibility in null-ls.

### Help

Yes

**Comment by keaising** [comment]
I have a try and it works as my wish. Here is what I have done:

1. Add a `postprocess` function after `vim.fn.writefile`: https://github.com/keaising/null-ls.nvim/commit/ef40cd2a304da245d698b5c09a6b7e4d4ad35b26
2. In my `null-ls` config, call `postprocess` function and format the config file with `jq`: [my null-ls config](https://github.com/keaising/dotfile/commit/c6f4b28ca7febba14fd307725f381a9c482edd14#diff-6859dc28992f2f89ef71e0980ec34d0670eeacbe1e863ffca84579d00eb49dc6R40-R43)

Would you consider merging this `postprocess` function into the main branch? I can make a merge request.

**Comment by jose-elias-alvarez** [comment]
I am not a fan of adding this as a specific option for the `cspell` source, which in my opinion has already grown too large - see my comments in #1412. I'd be fine with adding a general hook for all sources that accepts `params` as an argument, which would make it possible to do this.

**Comment by keaising** [comment]
Yeah, I agree with you, I think `postprocess` is a hook that meet your requirement: it can be called after the code action has been processed, all sources can benefit from this hook.

The reason I use `postprocess` as callback is that I found there is a `postprocess` in formatting and diagnostic, I think it might also be acceptable to add a similar `postprocess` in code_action for cspell, like this: https://github.com/keaising/null-ls.nvim/commit/ef40cd2a304da245d698b5c09a6b7e4d4ad35b26

If you have any idea about my implementation, please let me know.

**Comment by jose-elias-alvarez** [comment]
What I meant is that rather than adding a specific config option for the `cspell` source, we should have a generalized hook for all sources. The name `postprocess` is [already in use](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/01e035ac1c0ba1cc7a688110e3a9f939cab1f454/lua/null-ls/generators.lua#L100) but you can see how we could use similar logic to, say, add `on_success` and `on_error` hooks.

**Comment by keaising** [comment]
OK, how about change `postprocess` to `on_success`?

This is the update: https://github.com/jose-elias-alvarez/null-ls.nvim/commit/cc721916e31074908f7f844e6023db56fe90bbe1

And this is usage: https://github.com/keaising/dotfile/blob/master/nvim/.config/nvim/lua/plugins/null-ls.lua#L30-L33

**Comment by jose-elias-alvarez** [comment]
Apart from the name of the hook, what I mean is that this should be a general option for all sources, not just `cspell`. Your setup could then look like this:

```lua
null_ls.builtins.code_actions.cspell.with({
    config = {
        -- this is specific to cspell and so belongs under the config key
        find_json = function(_)
            return vim.fn.expand("~/.config/nvim/cspell.json")
        end,
    },
    -- works for all sources, not just cspell
    on_success = function(_)
        os.execute(
            "cat ~/.config/nvim/cspell.json | jq -S '.words |= sort' | tee ~/.config/nvim/cspell.json > /dev/null"
        )
    end,
})
```

The right place to make this change would be in [`generators.lua`](https://github.com/jose-elias-alvarez/null-ls.nvim/blob/01e035ac1c0ba1cc7a688110e3a9f939cab1f454/lua/null-ls/generators.lua). You can see the `postprocess` code I linked above - the implementation here should be similar.

**Comment by keaising** [comment]
I try to add a `on_success` like `postprocess`, you can find my commit at https://github.com/jose-elias-alvarez/null-ls.nvim/commit/660266f5779fe16232edd29c7819ea4cf3c7b0eb, but it doesn't work.

I think the reason is that code action has it own logic here: https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/lua/null-ls/code-actions.lua#L48-L51 

When I try to run `on_success` after the `s.run_action` in `code_actions.lua`, I can not find the `on_success` in `handler` function, maybe it is not a good idea to run `on_success` in `code_actions.lua`. 

Could you please give me some more information about what I should do to add this `on_success`?

**Comment by jose-elias-alvarez** [comment]
Hmm, that's a great point that I hadn't considered - my idea to run the hook after the generator returns results won't work, since that's not when code actions are actually applied. There is likely a better general purpose solution here, but if you want to open a PR with your changes for `cspell`, feel free.

**Comment by keaising** [comment]
I agree with you that cspell has too many unique options, so I think if you have a plan to add a common `on_success` to all sources, I can wait for it, but if you don't have that plan, I will open a PR with a unique `on_success` for `cspell`

**Comment by jose-elias-alvarez** [comment]
I'm not planning on doing so (and for the reasons you noted, we'd need something separate for code actions, anyways). Feel free to open a `cspell` specific PR with the changes from the commits you linked to.
